datalad_next/patches/push_to_export_remote.py

```diff
@@ -0,0 +1,240 @@
+
+    for line in repo.call_git_items_(["log", "--pretty=%H %T"]):
+        commit_hash, treeish = line.split()
+        if treeish == export_entry["treeish"]:
+            return True
+    return False
+
+
+def _transfer_data(repo: AnnexRepo,
+                   ds: Dataset,
+                   target: str,
+                   content: Iterable,
+                   data: str,
+                   force: Optional[str],
+                   jobs: Optional[Union[str, int]],
+                   res_kwargs: Dict,
+                   got_path_arg: bool
+                   ) -> Generator:
+
+    target_uuid, remote_info = ([
+        (uuid, info) for uuid, info in repo.get_special_remotes().items()
+        if info.get("name") == target] or [(None, None)])[0]
+
+    if _is_export_remote(remote_info):
+        # TODO:
+        #  - check for configuration entries, e.g. what to export
+
+        lgr.debug("Exporting HEAD to a remote with exporttree == yes")
+
+        if ds.config.getbool('remote.{}'.format(target), 'annex-ignore', False):
+            lgr.debug(
+                "Target '%s' is set to annex-ignore, exclude from data-export.",
+                target)
+            return
+
+        if force not in ("all", "export"):
+            export_entry = _get_export_log_entry(repo, target_uuid)
+            if export_entry:
+                if export_entry["source-annex-uuid"] != repo.uuid:
+                    yield dict(
+                        **res_kwargs,
+                        status="error",
+                        message=f"refuse to export to {target}, because the "
+                                f"last known export came from another repo "
+                                f"({export_entry['source-annex-uuid']}). Use "
+                                f"--force=export to enforce the export anyway.")
+                    return
+                if not _is_valid_treeish(repo, export_entry):
+                    yield dict(
+                        **res_kwargs,
+                        status="error",
+                        message=f"refuse to export to {target}, because the "
+                                f"current state is not a fast-forward of the "
+                                f"last known exported state. Use "
+                                f"--force=export to enforce the export anyway.")
+                    return
+
+        credentials = _get_credentials(ds, remote_info)
+
+        # If we have credentials, check whether we require an environment patch
+        env_patch = {}
+        remote_type = remote_info.get("type")
+        if credentials and needs_specialremote_credential_envpatch(remote_type):
+            env_patch = get_specialremote_credential_envpatch(
+                remote_type,
+                credentials)
+
+        res_kwargs['target'] = target
+
+        with patch.dict('os.environ', env_patch):
+            try:
+                repo.call_git(["annex", "export", "HEAD", "--to", target])
```

Comment:
It would be ideal to call it with `--json --json-error-messages`. This will produce reports like

```
{"command":"export demo","success":true,"input":[],"error-messages":[],"file":".datalad/.gitattributes"}
{"command":"export demo","success":true,"input":[],"error-messages":[],"file":".datalad/config"}
{"command":"export demo","success":true,"input":[],"error-messages":[],"file":".gitattributes"}
```

they could be feed into `annexjson2result()` and give us copyto-like result reports.

Reply:
Thx, will do that

Reply:
It could even to --json-progress and we could use `_call_annex_records_items_()`

Reply:
Adressed as suggested in commit 14b1d978f6eaa32d90cd2fe0131f04d1d101d5f9.

The function `_transfer_data` now uses `_call_annex_record_items_()` and yields all results immediately.

Reply:
FTR: git-annex does not seem to distinguish a file that is already up-to-date, and one that actually needs copying. The only way to discover that seems to be to look for `unexport` messages:

Here only `myfile.txt` has changes since the last export

```json
{"command":"unexport fz-juelich.sciebo.de-storage","success":true,"input":[],"error-messages":[],"file":"myfile.txt"}
{"byte-progress":125,"action":{"command":"export fz-juelich.sciebo.de-storage","input":[],"file":".datalad/.gitattributes"},"total-size":125,"percent-progress":"100%"}
{"command":"export fz-juelich.sciebo.de-storage","success":true,"input":[],"error-messages":[],"file":".datalad/.gitattributes"}
{"byte-progress":63,"action":{"command":"export fz-juelich.sciebo.de-storage","input":[],"file":".datalad/config"},"total-size":63,"percent-progress":"100%"}
{"command":"export fz-juelich.sciebo.de-storage","success":true,"input":[],"error-messages":[],"file":".datalad/config"}
{"command":"export fz-juelich.sciebo.de-storage","success":true,"input":[],"error-messages":[],"file":".gitattributes"}
{"byte-progress":5,"action":{"command":"export fz-juelich.sciebo.de-storage","input":[],"file":"myfile.txt"},"total-size":5,"percent-progress":"100%"}
{"command":"export fz-juelich.sciebo.de-storage","success":true,"input":[],"error-messages":[],"file":"myfile.txt"}

```